### PR TITLE
New Feature: Private room

### PR DESCRIPTION
**Pull request by magickaichen**
-Separate the functionality for joining and creating room
-Users now is able to create both public and private room upon their
choice
-Detect duplicate room while user attempting to create a room with the
same name
-Catch errors while users attempting to join the room does not exist

### CODE REVIEW

**Comment by magickaichen** [comment]
I have resolve the conflicts but it still cannot be automatically merged, gonna wait till the code review ends and apply the manual merge

**Comment by aralph** [comment]
I will start the review, but
- reviewing the code is hard bcs when I use "Files changed", I see the diff between your old master and your new master after the PR. This means I see a lot of changes, including the ones from my PR which are already merged in the CodeMonkeys master...
- testing the page is hard since I have to checkout your repo and compare it with our current master. This means that it is not possible to use the same local repo and simply checkout another version. 
  These two issues do not appear if you develop new code on a branch of the CodeMonkey repo rather than developing code in your own forked/cloned repo. 

I am repeating this since the people who write code and use this alternative workflow might not be aware that this causes a lot of extra work for the people who review the code.

File: css/main.css [hunk 1]
```diff
@@ -70,12 +70,24 @@ button{
...
 	position: fixed;
-	bottom: 95px;
-	right: 40px;
+	bottom: 60px;
+	right: 30px;
 	cursor: pointer;
-	display: none;
 }
-#toTop .fa {margin-right: 5px;}
+
```
Comment (aralph) [inline css/main.css:77]: As explained, these are code changes from my PR which are already merged. 
This is obvious here, but if several people work on the same lines of code, reviewing becomes impossible...

File: index.html [hunk 2]
```diff
@@ -34,65 +34,215 @@ <h3 class="masthead-brand">resPOSTa!</h3>
...
+                            autofocus
+                            aria-describedby="sizing-addon1"
+                            placeholder="Room name">
+                            <a href="#" class="btn btn-lg btn-default"
+                            ng-hide="form.input.$error.required || form.input.$error.pattern"
+                            ng-click="join()"
+                            >Join</a>
+                        </form>
+                    </p>
+                    <div class="alert alert-warning" role="alert" ng-show="form.input.$error.required">
```
Comment (aralph) [inline index.html:54]: Before entering the first letter, this alert is already displayed. 
You should couple the display of this alert to the pressing of the 'Create a Room' button (6 lines later on line 60).
Create an issue for this. 

bildschirmfoto 2015-11-19 um 22 01 57

File: index.html [hunk 3]
```diff
@@ -34,65 +34,215 @@ <h3 class="masthead-brand">resPOSTa!</h3>
...
+                                    <form role="form" name="CreateRoom">
+                                        <div class="form-group">
+                                            <label for="usrname"><span class="glyphicon glyphicon-home"></span> Room Name</label>
+                                            <input type="text" class="form-control" name="usrname" placeholder="Your Room Name"
+                                                   ng-model="newRoomName.text" ng-pattern="newRoomName.word" required>
+                                            <p role="alert" ng-show="CreateRoom.usrname.$error.required">
+                                                <b style="color: red;">Room name is required!</b>
+                                            </p>
+                                            <p role="alert" ng-show="CreateRoom.usrname.$error.pattern">
+                                                <b style="color: orange;">Room name must be English letters!</b>
```
Comment (aralph) [inline index.html]: "Room name must only contain English letters and numbers. "
or 
"Room name must only contain alphanumeric characters."

Fix this.

File: index.html [hunk 4]
```diff
@@ -34,65 +34,215 @@ <h3 class="masthead-brand">resPOSTa!</h3>
...
+                                        </div>
+                                        <!--Will ask user to input password if the room is a private room-->
+                                        <div class="form-group" ng-show="checkBoxValue.isPrivate">
+                                            <label for="psw"><span class="glyphicon glyphicon-barcode"></span> Password</label>
+                                            <input type="text" class="form-control" name="psw" placeholder="Enter password" ng-model="PasswordBoxValue.Password" ng-pattern="PasswordBoxValue.checkValid" required>
+                                            <p role="alert" ng-show="CreateRoom.psw.$error.required">
+                                                <b style="color: red;">Password is required!</b>
+                                            </p>
+                                            <p role="alert" ng-show="CreateRoom.psw.$error.pattern">
+                                                <b style="color: orange;">Password must be English letters!</b>
```
Comment (aralph) [inline index.html]: dito

File: index.html [hunk 5]
```diff
@@ -34,65 +34,215 @@ <h3 class="masthead-brand">resPOSTa!</h3>
...
+                            }
+                            // Room is public (NOT private), go to the room directly
+                            else{
+                                $window.location.href = 'question.html#/' + $scope.roomName.text;
+                            }
+                        }
+                    });
+                    // Room not found, return an alert
+                    if(!found){
+                        $window.alert("Room is not exist, click 'Create a New Room' to create a new room.");
```
Comment (aralph) [inline index.html:238]: "Room does not exist, click on 'Create A New Room' to create a new room."

File: js/controllers/todoCtrl.js [hunk 6]
```diff
@@ -29,28 +29,41 @@ function ($scope, $location, $firebaseArray, $sce, $localStorage, $window) {
...
 var firebaseURL = "https://cmkquestionsdb.firebaseio.com/";
+// Backup DB (TEST ONLY)
+//var firebaseURL = "https://questionstestdb.firebaseio.com/";
 
 // create variables for firebase DB
-$scope.roomId = roomId;
-var url = firebaseURL + roomId + "/questions/";
-var urlReplies = firebaseURL + roomId + "/replies/";
-var echoRef = new Firebase(url);
+
```
Comment (aralph) [inline js/controllers/todoCtrl.js:47]: It looks like you merged this part correctly. 
There were no issues while testing.

File: question.html [hunk 7]
```diff
@@ -19,6 +19,7 @@
```
Comment (aralph) [inline question.html:19]: There were no changes in this file, right?

**Comment by aralph** [comment]
Finished the code review. 
Only minor corrections needed. 
Mainly, fix the "Room name is required alert!" issue.
I want that our landing page looks neat. 

Otherwise, lgtm.
Try to do your manual merge.

File: question.html (hunk 7 above)
Comment (magickaichen) [inline question.html:19]: Yes, I didn't change anything, changes shown here is because my old master does not up to date.

**Comment by magickaichen** [comment]
Now I am trying to merge this Pull Request to our master branch since most of the minor issues has been fixed and Ralph says "lgtm" after the code review.
If anyone have something to say before I merge it please do so now.

**Pull request merged**